### DISCUSSION TITLE
Is it possible to perform a dynamic `rewrite` in `getStaticProps`?

### DISCUSSION THREAD

**Discussion by snelsi** [vercel/next.js#41178] [Help]
We are in the process of [Incrementally Adopting Next.js](https://nextjs.org/docs/migrating/incremental-adoption).

Some of our pages are rendered via Next.js and some are rendered via WordPress.
We are using one dynamic 'catch all' page + ISR to catch and render all routes. 

The actual list of pages:
- is stored in the WP database,
- changes often,
- has no distinguished URL pattern between new/old pages.

[Rewrite docs](https://nextjs.org/docs/api-reference/next.config.js/rewrites) gives an example of how to fetch and apply rewrites at build time. But we don't want to trigger an entire deployment process just to update rewrites.

Is it possible to perform a rewrite dynamically, inside `getStaticProps`?

**Comment by icyJoseph** [comment]
I guess you could implement middleware and check the incoming path against a list of valid paths in your DB, and based off that rewrite at the Edge. Note that if you don't have special setup for Edge, then `middleware` simply runs next to your server instance via `next start`.

You can return `redirect` from `getStaticProps`, but it ought to be routes created at runtime as part of ISR, and not during build time. https://nextjs.org/docs/api-reference/data-fetching/get-static-props#redirect -

However, a redirect would cause page unload. 

I am a bit short of time today, otherwise I'd do some testing and post some snippets. Please do follow up with more questions in case I am missing something, or have misunderstood your problem.

**Reply by snelsi** [reply]
You mean doing something like this?

```ts
import { NextResponse } from 'next/server'
import type { NextRequest } from 'next/server'

const fetchListOfWordpressPages = async (): Promise<string[]> => fetch('/api/getWordpressPages')

export async function middleware(request: NextRequest) {
  const listOfWordpressPages = await fetchListOfWordpressPages()

  const { pathname } = request.nextUrl

  if (listOfWordpressPages.includes(pathname)) {
    const rewriteURL = new URL(pathname, 'https://wordpress.my-site.com/')
    return NextResponse.rewrite(rewriteURL)
  }
}
```

Is there a way to cache the api result and not call it on every request?

**Reply by icyJoseph** [reply]
I think you should just reach for the WP db from your edge lambda, no need to make a jump to `/api`.

I guess nothing is stopping your from creating an object/Map at the top of the module, and saving data there, because that's just how closures work, however, if deploying to the edge there might be multiple of such runtimes, and also how would you evict, revalidate? 

Though I would try to make sure I can reach a fast CDN instead, and limit how many request arrive to this middleware via a matcher. Of course, each business case requires different trade-offs.

What was your thought with regard to this when using `getStaticProps`? That one would've also had to check with an api right?

Also if your site is behind a CDN, perhaps you can handle caching there, if there's no personalised requests that is.

**Reply by snelsi** [reply]
We are fetching the page's data in `getStaticProps`, so I don't need an extra API call there to know if a page exists or not. 
If the page exists, but it's on WordPress - the API will contain a related attribute in the response.

For me `getStaticProps` seems like a perfect place to perform a rewrite, since it fits into the existing `notFound`/`redirect` way of handling things.

**Reply by icyJoseph** [reply]
Yeah exactly that's the beauty of data fetching functions. Unfortunately, the rewrite is not really a thing they do. I guess you could muddle this by returning a flag in your props and having forking logic inside the JSX, but that sounds leaky and very bad. 

I guess, you could implement an endpoint (as a separate service) that responds with 204 if the page is a go, otherwise 404, and from the middleware you control where to rewrite that request. That way you skip having to serialise all of the available pages on every single request. A Redis DB could handle this. This is also a bit leaky though, but it'd get the job done, and cost a bit more perhaps?

Sorry I can't help mate.
